### PR TITLE
Add GeoTile and GeoHash Grid aggregations on GeoShapes.

### LINKED ISSUE #4072: [Feature] GeoTileGrid Aggregation on Geo Shape field
### Description
This issue tracks the implementation of GeoTileGrid Aggregation on Geo Shape field. 

### Tasks:
- [x] Refactor code of GeoTileGrid Aggregation on GeoPoint from server to modules/geo folder.
- [x] Implement the GeoHashGrid on GeoShape field
     - [x] Implement the code to find the intersection with the tiles and return the results in correct bucket.
     - [x] Integration Tests
     - [x] Unit Tests
### Dependency
https://github.com/opensearch-project/OpenSearch/issues/4049

### Useful Links
1. RFC for All Aggregations on GeoShape: https://github.com/opensearch-project/geospatial/issues/84
2. Low Level Implementation Proposal: https://github.com/opensearch-project/geospatial/issues/92
3. Enabling Doc values for GeoShape: https://github.com/opensearch-project/OpenSearch/issues/4049
4. All Aggregation Implementation issue: https://github.com/opensearch-project/geospatial/issues/104
5. https://github.com/opensearch-project/geospatial/issues/195

### PR DESCRIPTION
**Pull request by navneet1v**
### Description
1. Add GeoTile and GeoHash Grid aggregations on GeoShapes.
2. Added IT test for the GeoTile Aggregation.
3. Added IT and Unit tests for new GeoShape Aggregations.

### Issues Resolved
#4072 , #4071

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/GeoModulePluginIntegTestCase.java [hunk 1]
```diff
@@ -24,6 +26,8 @@ public abstract class GeoModulePluginIntegTestCase extends OpenSearchIntegTestCa
 
     protected static final double GEOHASH_TOLERANCE = 1E-5D;
 
+    protected static final WellKnownText WKT = new WellKnownText(true, new StandardValidator(true));
```
Comment (andrross) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/GeoModulePluginIntegTestCase.java:29]: Any reason to prefer putting this constant here, as opposed to making it a private constant in the one file that it is used in?

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoHashGridIT.java [hunk 2]
```diff
@@ -34,135 +34,67 @@
...
-            source = source.field("location", latLon);
-        }
-        source = source.endObject();
-        return client().prepareIndex(index).setSource(source);
-    }
-
-    private static IndexRequestBuilder indexCity(String index, String name, String latLon) throws Exception {
-        return indexCity(index, name, Arrays.<String>asList(latLon));
-    }
+    private static String SMALLEST_GEO_HASH = null;
```
Comment (andrross) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoHashGridIT.java:79]: Please be consistent with naming. This should probably be camelCase like the other mutable static fields here.

**Review by andrross — COMMENTED** [review]
Couple minor comments, but as I'm not a subject matter expert here I unfortunately can't really give a good review on the core of this change.

File: modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java [hunk 3]
```diff
@@ -105,23 +102,4 @@ public List<CompositeAggregationSpec> getCompositeAggregations() {
...
-
-    /**
-     * Registering the GeoBounds Aggregation on the GeoShape Field. This function allows plugins to register new
-     * aggregations using aggregation names that are already defined in Core, as long as the new aggregations target
-     * different ValuesSourceTypes.
-     *
-     * @return A list of the new registrar functions
-     */
-    @Override
-    public List<Consumer<ValuesSourceRegistry.Builder>> getAggregationExtentions() {
```
Comment (reta) [inline modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java:117]: Could you please clarify why this registration has been removed? I do not 100% understand the consequences yet, but the comment documents the intention pretty well why it is needed: `GEO_SHAPE` is defined in core.

File: modules/geo/src/main/java/org/opensearch/geo/search/aggregations/bucket/geogrid/util/GeoShapeHashUtil.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+     * @return {@link List} containing encoded {@link Long} values
+     */
+    public static List<Long> encodeShape(final GeoShapeDocValue geoShapeDocValue, final int precision) {
+        final List<Long> encodedValues = new ArrayList<>();
+        final GeoShapeDocValue.BoundingRectangle boundingRectangle = geoShapeDocValue.getBoundingRectangle();
+        long topLeftGeoHash = Geohash.longEncode(boundingRectangle.getMinX(), boundingRectangle.getMaxY(), precision);
+        long topRightGeoHash = Geohash.longEncode(boundingRectangle.getMaxX(), boundingRectangle.getMaxY(), precision);
+        long bottomRightGeoHash = Geohash.longEncode(boundingRectangle.getMaxX(), boundingRectangle.getMinY(), precision);
+
+        // Now iterate these 4 coordinates.
```
Comment (reta) [inline modules/geo/src/main/java/org/opensearch/geo/search/aggregations/bucket/geogrid/util/GeoShapeHashUtil.java:39]: Out of curiosity, I would have expected to see `bottomLeftGeoHash` somewhere but there are only 3 coordinates, not 4, typo?

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/GeoModulePluginIntegTestCase.java (hunk 1 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/GeoModulePluginIntegTestCase.java:29]: It is being used across the IT(GeoHash and GeoTile) hence I put it in the base class, and I am thinking it is going to get used for other IT also in future.

File: modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java (hunk 3 above)
Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java:117]: It was removed because I move it to [GeoBoundsAggregatorFactory.java](https://github.com/opensearch-project/OpenSearch/pull/5589/files#diff-8fa6a1ff75628d10d0b4dacc71c37f8abed5daed9cdcc7f313b556cdbe070ff2). When I was working on GeoBounds Aggregation I was thinking that to add new aggregations I need to put it in getAggregationExtentions(). But while I am working on GeoTile and GeoHash I realized that these aggregations can be added in *AggregatorFactory.class . The PR contains that change. 

I hope this clarifies.

File: modules/geo/src/main/java/org/opensearch/geo/search/aggregations/bucket/geogrid/util/GeoShapeHashUtil.java (hunk 4 above)
Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/search/aggregations/bucket/geogrid/util/GeoShapeHashUtil.java:39]: Actually ```bottomLeftGeoHash``` variable is not required. So the way this logic works is we are starting from topLeft and for every row we move till right most cell. and the loop is breaking when we reach the bottomRight cell.

Seems like the comment is misleading. I am using 3 coordinates but saying iterate over 4. Let me change it.

Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/search/aggregations/bucket/geogrid/util/GeoShapeHashUtil.java]: I have removed that comment.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoHashGridIT.java (hunk 2 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoHashGridIT.java]: updated.

File: modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java (hunk 3 above)
Comment (reta) [reply modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java:117]: Thanks, yeah, I saw that, really unsure about the consequences but the change (`AggregatorFactory`) looks good

**Comment by reta** [comment]
@navneet1v LGTM! experts signoff would be valuable, @nknize could you take a look please?

File: modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java (hunk 3 above)
Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java:117]: I tested the changes via IT and also manual testing it works. Also, AggregatorFactory is used to register the aggregation. Given that this is same aggregation but on a different SourceValue type I don't see any issue. The only thing I can think of is, is this the right place from a code abstraction standpoint. I will wait for @nknize to comment on this.

Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/GeoModulePlugin.java:117]: But as per my understanding of the Aggregation framework, I think this looks good, as aggregations are present in same module/plugin. So we are good. getAggregationExtentions : function is provided to extend the aggregations by the plugin for a different Source value type but having same aggregation name defined in core. Doc link: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/SearchPlugin.java#L174-L176

**Comment by navneet1v** [comment]
Flaky test failure. Issue: https://github.com/opensearch-project/OpenSearch/issues/1567

> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/9114/
> * **CommitID:** [0ae5eac](https://github.com/opensearch-project/OpenSearch/commit/0ae5eacae9da49c7897751a2c28ccce7a548b7ec)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ */
+public abstract class AbstractBucketAggregationIntegTest extends GeoModulePluginIntegTestCase {
+
+    protected static final int MAX_PRECISION_FOR_GEO_SHAPES_AGG_TESTING = 4;
+
+    protected static final int NUM_DOCS = 100;
+
+    protected static final String GEO_SHAPE_INDEX_NAME = "geoshape_index";
+
+    // We don't want to generate this BB at random as it may lead to generation of very large or very small BB.
```
Comment (nknize) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java:36]: It's true we don't want the test to be too demanding but we also don't want to miss potential edge cases caught by randomization. Maybe add a new method `GeoTestUtil.randomBox(double radius)`?

```java
private static Rectangle randomRectangle(double radius) {
    double centerLat = GeoTestUtil.nextLatitude();
    double centerLon = GeoTestUtil.nextLongitude();
    Polygon p = GeoTestUtil.createRegularPolygon(centerLat, centerLon, radius, 4);
    return new Rectangle(p.minLat, p.maxLat, p.minLon, p.maxLon);
}
```

If I remember right, this test util may generate irregular line polygons, so you may have to take some inspiration from lucene's `GeoTestUtil#nextBoxInternal(boolean canCrossDateline)`

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoHashGridIT.java [hunk 6]
```diff
@@ -34,135 +34,67 @@
...
-            cities.add(indexCity("multi_valued_idx", Integer.toString(i), points));
-            for (String hash : geoHashes) {
-                multiValuedExpectedDocCountsForGeoHash.put(hash, multiValuedExpectedDocCountsForGeoHash.getOrDefault(hash, 0) + 1);
-            }
-        }
-        indexRandom(true, cities);
-
-        ensureSearchable();
+        prepareSingleValueGeoPointIndex(random);
+        prepareMultiValuedGeoPointIndex(random);
```
Comment (nknize) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoHashGridIT.java:85]: 👍🏻

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+
+    private static final String AGG_NAME = "geotilegrid";
+
+    private static final String GEO_SHAPE_INDEX_NAME = "geoshape_index";
+
+    private static final String GEO_SHAPE_FIELD_NAME = "location_geo_shape";
+
+    private final Version version = VersionUtils.randomIndexCompatibleVersion(random());
+
+    private static final Rectangle BOUNDING_RECTANGLE_FOR_GEO_SHAPES_AGG = new Rectangle(-4.1, 20.9, 21.9, -3.1);
```
Comment (nknize) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:59]: same as above.. abstracting a base GridIntegTest will allow us to reuse the logic, and extend for general XYShape in the future.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertSearchResponse;
+
+@OpenSearchIntegTestCase.SuiteScopeTestCase
+public class GeoTileGridIT extends AbstractBucketAggregationIntegTest {
```
Comment (nknize) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:47]: This duplicates logic w/ `GeoHashGridIT` so maybe we create a new Abstract class  (e.g.,  `AbstractGeoGridIntegTest.java`) that provides the common logic across the two? (e.g., `randomRectangle`, `prepareSingleValueGeoPointIndex`, `testGeoShapes`)?

**Review by nknize — CHANGES_REQUESTED** [review]
Nice! Implementation is looking clean. I left a request to strengthen the tests with better randomization using some lucene geo test utils. I'd like to see if we can stress this a little further than the static Rectangle that's being used; and I think we can do that w/o introducing too much more complexity.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java (hunk 7 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:59]: Yes sure will do that.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java (hunk 8 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:47]: Some of the common logic has already been moved to an abstract class AbstractBucketAggregationIntegTest.class. As this was a geo module hence I thought to name it like this.

Now if we see other logic which is prepareSingleValueGeoPointIndex, testGeoShapes we can abstract some logic out but it will lead to multiple override functions and building tests in generic fashion.

I am not sure for integration tests I want to do that. The reason is if tests start to fail it becomes hard to debug currently we will get exact names that are failing. I understand there is some duplication but it provides some a better debugging experience. 

Having said that I will try to move more generic functions out to already created abstract class.

Function like randomRectangle will go to [RandomGeoGeometryGenerator](https://github.com/opensearch-project/OpenSearch/blob/56162a9f25b32a4f084bc21107e63aecc702ed4e/modules/geo/src/test/java/org/opensearch/geo/tests/common/RandomGeoGeometryGenerator.java#L39) class, which already have all helper functions for generating the random shapes.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java (hunk 7 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:59]: Actually this code was already abstracted it just I was not using it for this test case. Will fix it.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java (hunk 5 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java:36]: @nknize I think generating a BBOX with a specified width is a good idea.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java (hunk 7 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:59]: The abstraction is already done via AbstractBucketAggregationIntegTest.class

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java (hunk 5 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java:36]: sure will do that. I used a little different logic to prepare the BBox with a specific radius.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java (hunk 7 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:59]: The abstraction is already done via AbstractBucketAggregationIntegTest.class

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java (hunk 8 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:47]: Let me abstract the logic for preparing the indexes for GeoShapes and GeoPoints. Other logic like creating rectangle will also be added.

Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java:35]: abstracted the code,

**Comment by navneet1v** [comment]
@nknize can you please take a look on the PR?

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+import static org.opensearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+
+/**
+ * This is the base class for all the Bucket Aggregation related integration tests. Use this class to add common
+ * methods which can be used across different bucket aggregations. If there is any common code that can be used
+ * across other integration test too then this is not the class. Use {@link GeoModulePluginIntegTestCase}
+ */
+public abstract class AbstractBucketAggregationIntegTest extends GeoModulePluginIntegTestCase {
```
Comment (nknize) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java:41]: nit: maybe `AbstractGeoBucketAggregationIntegTest` to avoid confusion w/ other tests like `AbstractBucketMetricsTestCase`?

**Review by nknize — APPROVED** [review]
Beautiful!!! One little nitpick.. can be done in a follow up PR. Thanks!

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java (hunk 9 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractBucketAggregationIntegTest.java]: I have fixed it. @nknize can you please merge in this PR after squashing the commits.

**Comment by navneet1v** [comment]
> 

This PR is for 2.7 release. Will add this PR in 2.x branch once 2.6 is released

**Comment by andrross** [comment]
> > 
> 
> This PR is for 2.7 release. Will add this PR in 2.x branch once 2.6 is released

@navneet1v Just curious, why can't you backport this now for the next release? Is the code here somehow incomplete?

**Comment by navneet1v** [comment]
The code is not incomplete. This feature has a UI part also which is not completed. Hence I am not merging it in 2.x Once it is merged in 2.x then it has to go in next release which is 2.6

**Comment by andrross** [comment]
I understand we might hold of on announcing the feature until the UI part is ready, but if the server portion is complete why not merge it when it is ready? In general early and continuous integration is better and it is one less thing to keep track of. It is probably not a big deal in this case but may be a useful example as a precedent.

**Comment by navneet1v** [comment]
Help me understand this, if a feature is required to be launched in 2.7 can i merge it early like before 2.6 launch in 2.x branch?

**Comment by andrross** [comment]
From a purely technical point of view, if a backward compatible change is committed to main then it should be backported to 2.x. The risk of not doing so is that if another change comes in that touches the same part of code as this change did, then it won't be able to cleanly backport because of the divergent branches.

All committed code should be releasable. I believe the code here is in fact releasable (technically), but is being held back for other reasons, and that is what I'm pushing on. As a thought experiment, if we decided to scrap the 2.6 release and instead release 3.0, what would you do? Would you revert this code?

**Comment by navneet1v** [comment]
> From a purely technical point of view, if a backward compatible change is committed to main then it should be backported to 2.x. The risk of not doing so is that if another change comes in that touches the same part of code as this change did, then it won't be able to cleanly backport because of the divergent branches.
> 
> 
> 

What you are saying about this is correct. If conflicting changes come it won't be a clean backport.

But here is my question with the current branching strategy if we are working on a change that requires multiple commits are we not supposed to merge in main? Where each commit can be built separately?

If I merge the changes for a release which is going to happen like 3 months down the road putting it in 2.x branch will make the cutting of 2.6 branches difficult. We need to revert some changes from 2.x which will add more burden on the release owner.

The strategy I am implementing here is keep on merging changes in main branch and only backport a feature when all other parts are fully ready.

If let's say I merge a big feature in 1 single commit, then approving that PR will be very difficult as it might contain like 100 files in it. I also cannot create a feature branch where I can put all the approved commits and then finally merge in main and 2.x because I don't have permission to create feature branch or keep it in sync with main.


On scrapping of all next 2.x releases and releasing 3.0 using main case, if that happens and we cut a 3.x branch. I am hoping maintainers will provide enough time to revert the commits which are not intended to be released. 

Moreover, all this is my understanding based on what we do in our other repos that follow same branching strategy.

I am adding @dblock to comment here. As he originally proposed the branching strategy. He might have some thoughts here.

**Comment by andrross** [comment]
To be clear, the alternative I'm advocating here is that we should backport this PR and these server-side changes should be available to users in 2.6. Why is this not acceptable?

I am absolutely not advocating for developing a feature in a single big commit. But the incremental commits along the way should be releasable.
